apis/oam/v1alpha2/types.go

```diff
@@ -0,0 +1,857 @@
+	// Arguments to be passed to the command run by this container.
+	// +optional
+	Arguments []string `json:"args,omitempty"`
+
+	// Environment variables that should be set within this container.
+	// +optional
+	Environment []ContainerEnvVar `json:"env,omitempty"`
+
+	// ConfigFiles that should be written within this container.
+	// +optional
+	ConfigFiles []ContainerConfigFile `json:"config,omitempty"`
+
+	// Ports exposed by this container.
+	// +optional
+	Ports []ContainerPort `json:"ports,omitempty"`
+
+	// A LivenessProbe assesses whether this container is alive. Containers that
+	// fail liveness probes will be restarted.
+	// +optional
+	LivenessProbe *ContainerHealthProbe `json:"livenessProbe,omitempty"`
+
+	// A ReadinessProbe assesses whether this container is ready to serve
+	// requests. Containers that fail readiness probes will be withdrawn from
+	// service.
+	// +optional
+	ReadinessProbe *ContainerHealthProbe `json:"readiessProbe,omitempty"`
+
+	// TODO(negz): Ideally the key within this secret would be configurable, but
+	// the current OAM spec allows only a secret name.
+
+	// ImagePullSecret specifies the name of a Secret from which the
+	// credentials required to pull this container's image can be loaded.
+	// +optional
+	ImagePullSecret *string `json:"imagePullSecret,omitempty"`
+}
+
+// A ContainerisedWorkloadSpec defines the desired state of a
+// ContainerisedWorkload.
+type ContainerisedWorkloadSpec struct {
+	// OperatingSystem required by this workload.
+	// +kubebuilder:validation:Enum=linux;windows
+	// +optional
+	OperatingSystem *OperatingSystem `json:"osType,omitempty"`
+
+	// CPUArchitecture required by this workload.
+	// +kubebuilder:validation:Enum=i386;amd64;arm;arm64
+	// +optional
+	CPUArchitecture *CPUArchitecture `json:"arch,omitempty"`
+
+	// Containers of which this workload consists.
+	Containers []Container `json:"containers"`
+}
+
+// A ResourceReference refers to an resource managed by an OAM resource.
+type ResourceReference struct {
+	// APIVersion of the referenced resource.
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of the referenced resource.
+	Kind string `json:"kind"`
+
+	// Name of the referenced resource.
+	Name string `json:"name"`
+
+	// UID of the referenced resource.
+	// +optional
+	UID *types.UID `json:"uid,omitempty"`
+}
+
+// A ContainerisedWorkloadStatus represents the observed state of a
+// ContainerisedWorkload.
+type ContainerisedWorkloadStatus struct {
+	runtimev1alpha1.ConditionedStatus `json:",inline"`
+
+	// Resources managed by this containerised workload.
+	Resources []ResourceReference `json:"resources,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// A Server is a long-running, scalable workload that has a network endpoint
+// with a stable name.
+// +kubebuilder:subresource:status
+type Server struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ContainerisedWorkloadSpec   `json:"spec,omitempty"`
```

Comment:
Since a component only inline a CR for this CRD,  the OAM controller won't even know if the component is referring to a "server" workload or a "singletonServer" anymore.  I wonder how can this work?

Reply:
There might be another issue. The ContainerisedWorkloadSpec doesn't have GVK in it. I guess "server" is the schema instead of a workload then? How does this work? Do we actually have a "server" workload defined (which has a definitionRef in it) somewhere?

Reply:
This changed in my latest push of this PR because we agreed to use a single workload type named `ContainerisedWorkload`. At the time you made this comment there were multiple workload kinds - e.g. `Server`, `SingletonServer`, etc that all happened to use the same struct for their spec and status. Does that make sense?